accounts-db/src/accounts_index.rs

```diff
@@ -1586,24 +1588,55 @@ impl<T: IndexValue, U: DiskIndexValue + From<T> + Into<T>> AccountsIndex<T, U> {
         self.account_maps.len()
     }
 
+    /// remove the earlier instances of each pubkey when the pubkey exists later in the `Vec`.
+    /// Could also be done with HashSet.
+    /// Returns `HashSet` of duplicate pubkeys.
+    fn remove_older_duplicate_pubkeys(
+        items: &mut Vec<(Pubkey, (Slot, T))>,
+    ) -> Option<Vec<(Pubkey, (Slot, T))>> {
+        if items.len() < 2 {
+            return None;
+        }
+        // stable sort by pubkey.
+        // Earlier entries are overwritten by later entries
+        items.sort_by(|a, b| a.0.cmp(&b.0));
+        let mut duplicates = None::<Vec<(Pubkey, (Slot, T))>>;
+        let mut i = 0;
+        while i < items.len().saturating_sub(1) {
```

Comment:
The while loop complexity- Complexity O(M*N). M - number of duplicates; N - Number of elements in the vec, 
can be optimized to O(N) with the following code.

```
let mut write = 1;  // pointer to the write
let mut curr = 1;   // pointer to current item
let mut prev_key = items[0].0;

while curr < items.len() {
  let curr_item = *item[curr];

  if curr_item.0 == prev_key {    
     let mut duplicates_insert = duplicates.unwrap_or_default();
     duplicates_insert.push(curr_item);
     duplicates = Some(duplicates_insert);
     curr += 1;
  }
  else {
    if (curr > write ) {
        items[write] = curr_item;
    }
    curr += 1;
    write += 1;
    prev_key = curr_item.0;
  }
}

items.truncate(write);
duplicates
```

Reply:
We expect m to be 0 always. This is purely defensive code.
But what you wrote looks correct and is faster if m > 0.

Reply:
yeah. we should iterate from the end. 

anyway, it doesn't matter.